backend/src/connector/client.rs

```diff
@@ -0,0 +1,429 @@
+            cred_marker: PhantomData,
+        })
+    }
+}
+
+impl<S, C, Cr> Client<S, C, Cr>
+where
+    Cr: Credentials + Clone,
+    S: Service<(Request<Body>, C), Response = Response<Body>> + Clone + Sync + Send + 'static,
+    S::Future: Send + 'static,
+    S::Error: Into<super::api::ServiceError> + std::fmt::Display + error_stack::Context,
+    C: Clone + Send + Sync + Has<XSpanIdString> + Has<Option<Authorization<Cr>>>,
+{
+    fn form_request(
+        &self,
+        endpoint: &str,
+        method: Method,
+        context: &C,
+    ) -> Result<Request<Body>, ClientError> {
+        let uri = format!("{}{endpoint}", self.base_path);
+
+        let uri = Uri::from_str(&uri).change_context(ClientError::BadUri)?;
+        let mut request = Request::builder()
+            .method(method)
+            .uri(uri)
+            .body(Body::empty())
+            .change_context(ClientError::CantFormRequest)?;
+        let xspan = Has::<XSpanIdString>::get(context);
+        let header =
+            HeaderValue::from_str(&xspan.0).change_context(ClientError::CantFormRequest)?;
+
+        request
+            .headers_mut()
+            .insert(HeaderName::from_static("x-span-id"), header);
+
+        let auth_data = Has::<Option<Authorization<Cr>>>::get(context);
+        if let Some(auth) = auth_data {
+            request
+                .headers_mut()
+                .typed_insert::<Authorization<Cr>>(auth.clone());
+        }
+
+        Ok(request)
+    }
+
+    async fn handle_response_json<R: for<'a> Deserialize<'a>, T>(
+        &self,
+        response: Response<Body>,
+        body_handler: impl Fn(R) -> T + Send,
+    ) -> Result<T, APIError> {
+        let body = response.into_body();
+        let body = body
+            .into_raw()
+            .await
+            .change_context(APIError::ResponseError)?;
+        let body = std::str::from_utf8(&body).change_context(APIError::ResponseError)?;
+
+        let body = serde_json::from_str::<R>(body)
+            .change_context(APIError::ResponseError)
+            .attach_printable("Response body did not match the schema")?;
+
+        Ok(body_handler(body))
+    }
+}
+impl<S, C, Cr> Client<S, C, Cr>
+where
+    Cr: Credentials + Clone,
+    S: Service<(Request<Body>, C), Response = Response<Body>> + Clone + Sync + Send + 'static,
+    S::Future: Send + 'static,
+    S::Error: Into<super::api::ServiceError> + std::fmt::Display + error_stack::Context,
+    C: Clone + Send + Sync + Has<RequestTimeout>,
+{
+    async fn call(&self, req: Request<Body>, cx: &C) -> Result<Response<Body>, APIError> {
+        let timeout = Has::<RequestTimeout>::get(cx);
+        tokio::time::timeout(
+            timeout.clone().into_inner(),
+            self.client_service.clone().call((req, cx.clone())),
+        )
+        .await
+        .change_context(APIError::RequestFailed)
+        .attach_printable("No Response received")?
+        .change_context(APIError::RequestFailed)
```

Comment:
Second call to `change_context(APIError::RequestFailed)`

Reply:
This intended since there are two distinct errors: the first one from tokio (request timed out) and the second one from hyper